cadence/contracts/services/AxelarGovernanceService.cdc

```diff
@@ -0,0 +1,296 @@
+        target: Address,
+        proposedCode: String,
+        eta: UInt64
+    )
+
+    pub event ProposalCancelled(
+        proposalHash: String,
+        target: Address,
+        proposedCode: String,
+        eta: UInt64
+    )
+
+    pub event ProposalExecuted(
+        proposalHash: String,
+        target: Address,
+        proposedCode: String,
+        timestamp: UInt64
+    )
+
+
+    init(gateway: Address, governanceChain: String, governanceAddress: String, minimumTimeDelay: UInt64){
+        if (governanceChain.length == 0 || governanceAddress.length == 0){
+            panic("InvalidAddress")
+        }
+
+
+        self.gateway = gateway
+        self.governanceChain = governanceChain
+        self.governanceAddress = governanceAddress
+        self.minimumTimeDelay = minimumTimeDelay
+        self.SELECTOR_SCHEDULE_PROPOSAL = Crypto.hash("scheduleProposal".utf8, algorithm: HashAlgorithm.KECCAK_256)
+        self.SELECTOR_CANCEL_PROPOSAL = Crypto.hash("cancelProposal".utf8, algorithm: HashAlgorithm.KECCAK_256)
+        self.inboxAccountCapabilityNamePrefix = "GovernanceUpdaterInbox_"
+        self.prefixAuthCapabilityName = "GovernanceUpdaterCapability_"
+        self.UpdaterContractAccountPrivatePath = PrivatePath(identifier: "UpdaterContractAccount_".concat(self.account.address.toString()))!
+        self.updaters <- {}
+        self.proposals <- {}
+
+        let axelarExecutabe <- create ExecutabeResource()
+        self.account.save(<-axelarExecutabe, to: /storage/AxelarExecutable)
+    }
+
+    
+    //Get estimated execution time for proposal
+    access(all) fun getProposalEta(proposedCode: String, target: Address): UInt64{
+        let proposalHash: String = String.encodeHex(self.createProposalHash(proposedCode: proposedCode, target: target))
+        return self.proposals[proposalHash]?.getTimeToExecute()!
+    }
+
+    //Execute Scheduled Proposal
+    access(all) fun executeProposal(proposedCode: String, target: Address){
+        let proposalHash: String = String.encodeHex(self.createProposalHash(proposedCode: proposedCode, target: target))
+        //check for time left in propsoal
+        if(self.proposals[proposalHash]?.getTimeToExecute()! <= UInt64(getCurrentBlock().timestamp)){
+            //Execute Proposal
+            self.proposals[proposalHash]?.execute()
+
+            emit ProposalExecuted(
+                proposalHash: proposalHash,
+                target: target,
+                proposedCode: proposedCode,
+                timestamp: UInt64(getCurrentBlock().timestamp)
+            )
+
+            destroy <- self.proposals.remove(key: proposalHash)
+        } else { 
+            panic("ProposalNotReady")
+        }
+    }
+
+    access(all) resource ExecutabeResource: AxelarGateway.Executable, AxelarGateway.SenderIdentity{
+        access(all) fun executeApp(commandResource: &AxelarGateway.CGPCommand, sourceChain: String, sourceAddress: String, payload: [UInt8]): AxelarGateway.ExecutionStatus{
```

Comment:
I believe the validation of a given command is handled at `AxelarGateway.execute`, is that correct?

Reply:
Correct!